### ISSUE TITLE
SuperClusterAlgorithm: change in marker position does not update the clustering

### ISSUE THREAD

**Issue by Nivani** [googlemaps/js-markerclusterer#606]
#### Environment details

OS type and version: /
Library version: 2.0.15

#### Steps to reproduce

  1. Setup a map with markers and marker clusterer (eg. the sample found at https://developers.google.com/maps/documentation/javascript/marker-clustering)
  2. Update the position of a marker using `markerObject.setPosition(new google.maps.LatLng(lat, lng));` in a way that should change the clustering
  3. Trigger a rerender on the clusterer: `clusterer.render()` -> NOK, expected the clustering to update

I did some debugging and found that the issue is here: https://github.com/googlemaps/js-markerclusterer/blob/2ac332afb468a2346ee1a56976fa2bd2c8e89291/src/algorithms/supercluster.ts#L52-L59

Because `this.markers` and `input.markers` contains the same marker instances, `equal(input.markers, this.markers)` is always `true` even though a copy of the array is made. The change detection will only detect when markers are added or removed.

**Comment by wangela** [comment]
If you would like to upvote the priority of this issue, please comment below or react with :+1: so we can see what is popular when we triage.
--------

@Nivani Thank you for opening this issue. 🙏
Please check out these other resources that might help you get to a resolution in the meantime:

- [Check the issue tracker](https://developers.google.com/maps/support/#issue_tracker) - bugs and feature requests for Google Maps Platform APIs and SDKs
- [Open a support case](https://developers.google.com/maps/support#contact-maps-support) - Get 1:1 support in Cloud Console.
- [Discord](https://discord.gg/fHTDmZe9En) - chat with other developers
- [StackOverflow](http://stackoverflow.com/questions/tagged/google-maps) - use the `google-maps` tag


_This is an automated message, feel free to ignore._

**Comment by Nivani** [comment]
A possible solution would be to save a copy of the relevant marker information in `this.markers` instead of the whole marker object.

For example (untested code):
```Javascript
public calculate(input: AlgorithmInput): AlgorithmOutput {
    let changed = false;

    const relevantMarkerInformation = input.markers.map((m) => ({
        lat: m.position.lat,
        lng: m.position.lng,
    }); 
    if (!equal(relevantMarkerInformation, this.markers)) {
      changed = true;
      this.markers = relevantMarkerInformation;

```

**Comment by zuizuihao** [comment]
Same problem happens on me. When I change the visible field of the markers, the clusters are not updated.

**Comment by vicb** [comment]
That is a limitation of the current library.

The clusters are not re-computed when the marker positions change.

@Nivani proposal could work but it will use more CPU (=lag) and more memory.

Maybe what could be done is to add way to force the algos to recompute the clusters.

It could be:
- adding a `force` parameter to `MarkerClusterer.render` (default to `false`) - or a `reset` method,
- passing this `force` parameter to the algos somehow - either in `calculate` or by adding a `reset` method to the interface.

Users would be responsible to call `render(true)` / `reset` when they move markers. I think it could be more efficient than trying to detect when it occurs.

@amuramoto what do you think?

**Comment by amuramoto** [comment]
Agreed that a reset method would be a good approach. Wonder if there's a straightforward way to provide it some identifier for the marker(s) that have changed so that just the relevant clusters can be updated efficiently.

**Comment by Nivani** [comment]
A reset method would work in our case.
